src/runtime/persistence/persistence.go

```diff
@@ -0,0 +1,70 @@
+// Package persistence provides persistent storage across low-power
+// and reset states.
+//
+// The amount of space available is under the control of the application
+// logic.
+//
+// For SRAM and built-in flash a single allocation can be made per
+// application (usually in the 'main' code) per persistence type.  I.e.
+// RAM and Flash are allocated separately so one application may have
+// both a RAM region and a Flash region, but not more than one of each.
+//
+// Libraries should NOT allocate persistence regions, but should accept
+// persistence regions passed in - and may indicate how much space is
+// required using public constants (for example).
+//
+// The built-in regions exposed are deliberately raw - there is no
+// native functionality to detect an uninitialized region, measures
+// to detect corruption, or functionality to implement transactions.
+// These are likely very application specific and left to the consumer
+// and/or external libraries.
+//
+package persistence
+
+// Region enables access to a persistence region.
+//
+// Alternate implementations of Region provide access to different
+// region types (RAM, SRAM, others).  External implementations
+// could provide access to battery backed RAM, or SPI flash
+// consistent with this interface.
+//
+type Region interface {
+	// Size gets the size of the persistence area (in bytes)
+	Size() int64
+
+	// ReadAt reads from the persistence area, it is signature
+	// compatible with Go's `io.ReaderAt` interface
+	ReadAt(dest []byte, offset int64) (n int, err error)
+
+	// WriterAt reads from the persistence area, it is signature
+	// compatible with Go's `io.WriterAt` interface
+	WriteAt(src []byte, offset int64) (n int, err error)
+
+	// SubAllocation creates a new persistence object that
+	// represents a sub-allocation of the main persistence
+	// area
+	SubAllocation(offset int, len int) (p Region, err error)
```

Comment:
What is the imagined use-case of this? Seems like it could be left out (for now) to simplify the interface until implementations have been fleshed out and tested. 

It seems like more of a convenience that isn't truly essential for this initial eval.

Reply:
I was thinking along the lines of slices in Go - being able to divide the space up, so parts can be handed off to different libraries without risk of bounds overflow.